### ISSUE TITLE
Design System: Audit Dropdown Element

### ISSUE THREAD

**Issue by adam-abundis** [hackforla/website#2286]
### Dependency
- [x] #3635 
- [ ] Waiting for wiki to be moved to its new repo and be ready for new content

### Overview
As a developer, we need to audit the HTML, CSS, and JavaScript for the dropdown element in order to ensure best practices are being applied. 

### Action Items
- [x] Find all instances of the HTML, CSS, or (inclusive) JS code for the dropdown element from all pages (one can be found on the about page) as of issue creation.
  - [x] Document all instances on a spreadsheet in the Audits folder in our drive with columns for Webpage, Dropdown Name, Permalink to Lines in the Code, and any other columns you think would be helpful
  - [x] Once the spreadsheet is created, add a link to the spreadsheet under the "Resources/Instructions" section below 
- [x] Research online on what the standards are in terms of HTML tags, accessibility, etc for this component.

### Resources/Instructions
<details>
  <summary>Click here to see example of the dropdown element</summary>
<br>
<img src="https://user-images.githubusercontent.com/21162229/133937942-0fd1fb27-834e-4dc6-a105-224effeb44a2.gif" />
</details>

- [Audit folder](https://drive.google.com/drive/folders/1yxGQ64wP5aN-4-0UC1RF0nYBhNaRCNsA)
- [Spreadsheet](https://docs.google.com/spreadsheets/d/1Ss_l-_xANUsfcpTMdR-skdKiTEgZidL3UnenBupXkZc/edit#gid=774837932)
- [Research](https://docs.google.com/document/d/1txTPpHkNC3UNkgoUbgxzASjalHbSjJEkdCDsdyXiaqY/edit#)

**Comment by d-perez-8** [comment]
Availability: Tuesday - Thursday
ETA: End of day Wednesday

**Comment by d-perez-8** [comment]
**Blocker**: I have created the spreadsheet [here](https://docs.google.com/spreadsheets/d/1Ss_l-_xANUsfcpTMdR-skdKiTEgZidL3UnenBupXkZc/edit#gid=774837932) from a template already existing in the [audit folder](https://drive.google.com/drive/folders/1yxGQ64wP5aN-4-0UC1RF0nYBhNaRCNsA), and I would confirmation that this is what was need. Thank you!

**Comment by d-perez-8** [comment]
Progress: I have learned how to create permalink to add to the spreadsheet.
Blockers: No blockers at the moment.
Availability: Will have time today and tomorrow, and I expect to finish it by Monday

**Comment by kathrynsilvaconway** [comment]
@d-perez-8 Hi, just checking in to see if you need any assistance with this issue. Be sure to comment your ETA and Availability for the week, and please do reach out on Slack if you are having a blocker.

**Comment by d-perez-8** [comment]
@kathrynsilvaconway Hi Kathryn. I do not need assistance yet, but I am still working on it! ETA by EOD, and I'm available all day today.

**Comment by d-perez-8** [comment]
I am finished with this issue and looking for it to get reviewed. Here is the [Spreadsheet](https://docs.google.com/spreadsheets/d/1Ss_l-_xANUsfcpTMdR-skdKiTEgZidL3UnenBupXkZc/edit#gid=774837932) @hackforla/website-merge

**Comment by arpitapandya** [comment]
> I am finished with this issue and looking for it to get reviewed. Here is the [Spreadsheet](https://docs.google.com/spreadsheets/d/1Ss_l-_xANUsfcpTMdR-skdKiTEgZidL3UnenBupXkZc/edit#gid=774837932) @hackforla/website-merge

Will review by EOD 2022-10-10
Availability : 4hrs

**Comment by blulady** [comment]
@d-perez-8 
Thanks for working on this Danny!! I noticed this instruction "Research online on what the standards are in terms of HTML tags, accessibility, etc for this component." I was wondering if I missed where you put the research?

**Comment by blulady** [comment]
Do the drop downs on the https://www.hackforla.org/ and https://www.hackforla.org/#press count too?

**Comment by d-perez-8** [comment]
@blulady Hey Sarah. I attached the research google doc on google sheets, and I have included it in the folder.

The drop-downs on the homepage are the same ones that are on the projects page. I included the homepage in the same line as the projects page on the google sheet. 

I did not see drop-downs on [press](https://www.hackforla.org/#press). I checked the desktop and mobile views, and it appears as a list to me.

**Comment by d-perez-8** [comment]
@hackforla/website-merge Can I get a review on this please 🙂

**Comment by blulady** [comment]
> @blulady Hey Sarah. I attached the research google doc on google sheets, and I have included it in the folder.
> 
> The drop-downs on the homepage are the same ones that are on the projects page. I included the homepage in the same line as the projects page on the google sheet.
> 
> I did not see drop-downs on [press](https://www.hackforla.org/#press). I checked the desktop and mobile views, and it appears as a list to me.

The one on press is the same drop down that is on the home page and projects page.

**Comment by blulady** [comment]
Thanks for attaching the Research document. I took a look at it and It would be helpful to have a bit more discussion on the accessibility aspect. I saw that you attached a really awesome article. I was hoping you could maybe include some of the key points from that article about accessibility and tags. And maybe a paragraph at the end summarizes your research findings stating which tags are preferred.

Things you might consider including in the research/research summary (obviously paraphrase or otherwise shorten this to include it in a research document):
[W3](https://www.w3.org/WAI/tutorials/menus/flyout/#flyoutnavkbbtn) 
"People with reduced dexterity, such as tremors, often have trouble operating fly-out menus. For some, it might be impossible. Make sure to provide other ways to reach the submenu items, for example by repeating them on the page of the parent menu item." 
"In addition, scripting is used to slightly delay the immediate closing of submenu items when the mouse leaves the area. This delay makes it easier to use the menu when navigation by a mouse is not very precise."

[Mozilla](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/select)
Here you discuss the proper use of the select element. It would be helpful to also mention how hard the select element is to style with css for accessibility. 
"The <select> element is notoriously difficult to style productively with CSS... consider using a library with good facilities for styling form widgets, or try rolling your own dropdown menu using non-semantic elements, JavaScript, and [WAI-ARIA](https://developer.mozilla.org/en-US/docs/Learn/Accessibility/WAI-ARIA_basics) to provide semantics."
Including a brief discussion of when you should look to using [WAI-ARIA](https://developer.mozilla.org/en-US/docs/Learn/Accessibility/WAI-ARIA_basics) with a drop down menu.
https://developer.mozilla.org/en-US/docs/Learn/Accessibility/WAI-ARIA_basics

**Comment by blulady** [comment]
[Building Accessible Menu Systems](https://www.smashingmagazine.com/2017/11/building-accessible-menu-systems/)

**Comment by blulady** [comment]
[Building Accessible Menu Systems](https://www.smashingmagazine.com/2017/11/building-accessible-menu-systems/)
Some things you might include
"Navigation systems are really tables of content and this is how tables of content are structured. The only thing that really makes us think “menu” is the styling of the nested lists and the way they are revealed on hover or focus."
"That’s where some go wrong and start adding WAI-ARIA semantics: 
ARIA menus are not designated for navigation but for application behavior. Imagine the menu system for a desktop application.
The top-level link should be usable as a link, meaning it does not behave like a menu button."

"Navigation submenus (or "dropdowns" to some) work well with a mouse or by keyboard, but they’re not so hot when it comes to touch. When you press the top-level "Shop" link in our example for the first time, you are telling it to both open the submenu and follow the link."  This can be resolved by "Make sure each top-level destination page has a table of contents as an alternative to the submenu."

"A [true menu](https://www.smashingmagazine.com/2017/11/building-accessible-menu-systems/#the-menu-and-menuitem-roles) (in the WAI-ARIA sense) should identify itself as such using the menu role (for the container) and, typically, menuitem children" And why you shouldn't use them.

"And <select> elements are menus of sorts, with similar semantics to the button-triggered menu we shall soon be constructing. However, just as with the checkbox toggle button, we’re using an element associated with entering input"

**Comment by blulady** [comment]
Also a brief summary/discussion of [True Menus](https://www.smashingmagazine.com/2017/11/building-accessible-menu-systems/#true-menus) and [Keyboard And Focus Behavior](https://www.smashingmagazine.com/2017/11/building-accessible-menu-systems/#keyboard-and-focus-behavior) would provide a lot of clarity to whoever reads this Research and is perhaps the most important part to mention and reference.

**Comment by blulady** [comment]
@d-perez-8 Isn't feeling well and is also going on Vacation until November 10th so he won't be able to keep working on this issue.

**Comment by gstemmann** [comment]
i. Availability: Tuesday 6-9 pm
ii. ETA: 11/10/22

**Comment by gstemmann** [comment]
Progress: I have reviewed the past work and spreadsheets that were completed.
Blockers: No blockers at the moment.
Availability: Will have time today

**Comment by gstemmann** [comment]
Progress: I went over more expectations on the research document with some members of the merge team.
Blockers: No blockers
Availability: I have interviews and I'm stating a new job (unrelated to the interview) so I will not be active/available this week.

**Comment by gstemmann** [comment]
Progress: Still progressing through the research and writing up the document with a new format
Blockers: Moving and started new job along with family time through the holidays
Availability: Friday Sat 11a - 3pm
ETA: Dec 3

**Comment by gstemmann** [comment]
# Write up:
Summary:
Navigation Submenus can be great with mouse and keyboard. But with touchscreens and for accessibility purposes, it's better to have a single menu with all available options. So when you press the top-level ["Shop"](https://cloud.netlifyusercontent.com/assets/344dbf88-fdf9-42bb-adb4-46f01eedd629/82a634a6-cfbc-4e9b-a6ad-a79476025c7e/1-menu-buttons-large-opt.png) link in our example for the first time, you should get a list of all the submenus instead of following a link. Also, we may want every top-level destination page to have a 'table-of-contents' as an alternative to the submenu. ARIA rules will override HTML roles, so in order to avoid creating a true menu with clickable menu items, the menu is not opened and closed by a “true menu button”.

Remember that navigation systems are just tables of content. We think of 'menu' because the lists are nested and they appear on hover. However, for accessibility, it’s important to make sure that each menu item element does not have interactive descendants. Also, adding a slight delay instead of immediately closing submenus (fly-out menus/dropdowns) when your mouse leaves the area can also help when the mouse isn’t precisely hovering over the menu.

As for keyboard accessibility, the best thing to do is use the correct elements. `ENTER` and `SPACE` buttons will fire events with `<button>` elements properly. Signposts like search bars and lists also are great for keyboard users and can be implemented with WAI-ARIA. 
(WAI-ARIA also is best to be used to help with nested `div` along with CSS and JavaScript since they can get complex and screen readers have a tough time working out what each feature does.)


Recommendations (Based on Research Findings above):
- [ ] The `<select>` element can be difficult to style with CSS so consider using a library that works well for styling form widgets. Also, some libraries aren’t readable by screen readers at all. So it may be easier to create your own dropdown menus with non-semantic elements (`<nav>`, `<footer>`, etc.), JavaScript, and WAI-ARIA for semantics.
- [ ] Nested `<div>` along with CSS and JavaScript can get complex and native control is changed with JavaScript. So it’s better to use the `<a>` element.
- [ ] Adding a slight delay instead of immediately closing submenus when your mouse leaves the area can help when the mouse isn’t precisely hovering over the menu item. This is especially true for touchscreens. This can be done with scripting.



### Research links / Resources:
* [Building accessible menu systems](https://www.smashingmagazine.com/2017/11/building-accessible-menu-systems/)
* [Fly-out menus](https://www.w3.org/WAI/tutorials/menus/flyout/#flyoutnavkbbtn)
* [When to use WAI-ARIA](https://www.smashingmagazine.com/2017/11/building-accessible-menu-systems/)
* [Select Elements](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/select)

**WCAG Standards:**
* [Focus Appearance](https://www.w3.org/WAI/standards-guidelines/wcag/new-in-22/#guideline-24-navigable)
* [How to Meet WCAG (Quick Reference)](https://www.w3.org/WAI/WCAG21/quickref/)
* [Grouping related links using the nav element](https://www.w3.org/WAI/WCAG21/Techniques/html/H97)
    * [ARIA11 Technique - using ARIA landmarks](https://www.w3.org/WAI/WCAG21/Techniques/aria/ARIA11.html)
    * [ARIA Landmarks - General Principles](https://www.w3.org/WAI/ARIA/apg/example-index/landmarks/index.html)
* [Keyboard Accessible - sufficient techniques](https://www.w3.org/WAI/WCAG21/quickref/?showtechniques=211#principle2) 
    * [Using CSS when component receives focus](https://www.w3.org/WAI/WCAG21/Techniques/css/C15.html) - [Focus visible](https://www.w3.org/WAI/WCAG21/Understanding/focus-visible)
    * [Navigable Guideline 2.4](https://www.w3.org/WAI/standards-guidelines/wcag/new-in-22/)
    * [Focus Not Obscured](https://www.w3.org/WAI/standards-guidelines/wcag/new-in-22/#2412-focus-not-obscured-minimum-aa)
    * [Draft - Understanding Focus Not Obscured](https://www.w3.org/WAI/WCAG22/Understanding/focus-not-obscured-minimum)

**Comment by gstemmann** [comment]
Progress: currently waiting for approval from dev lead
Availability: Sunday 12p - 4p

**Comment by blulady** [comment]
Hey George!! This is really good,thorough, work. I appreciate you picking it up. I apologize for taking so long to get back to you about it. I have a few suggestions. Let me know if you think they make sense. I think we should also place [this content](https://github.com/hackforla/website/issues/2286#issuecomment-1328145114) into the [research google doc](https://docs.google.com/document/d/1txTPpHkNC3UNkgoUbgxzASjalHbSjJEkdCDsdyXiaqY/edit#) although it doesn't expressly say it in the issue because it's nice to have as much information in one place as possible (namely the audits folder). With that in mind, we might also add a link in the research to this issue. I think it would also be nice to have a link to the Shop example you cite in the first paragraph, third sentence.
Thank You!!

**Comment by gstemmann** [comment]
> Hey George!! This is really good,thorough, work. I appreciate you picking it up. I apologize for taking so long to get back to you about it. I have a few suggestions. Let me know if you think they make sense. I think we should also place [this content](https://github.com/hackforla/website/issues/2286#issuecomment-1328145114) into the [research google doc](https://docs.google.com/document/d/1txTPpHkNC3UNkgoUbgxzASjalHbSjJEkdCDsdyXiaqY/edit#) although it doesn't expressly say it in the issue because it's nice to have as much information in one place as possible (namely the audits folder). With that in mind, we might also add a link in the research to this issue. I think it would also be nice to have a link to the Shop example you cite in the first paragraph, third sentence. Thank You!!

Awesome, Sarah! Thank you for your feedback. I made the changes you recommended and I absolutely agree it'd be helpful to have as much information as possible in the Audits folder. Please let me know if you have any other thoughts.

**Comment by gstemmann** [comment]
Progress: Just made some changes recommended by the merge team
Blockers: No blockers currently
Availability: 6 hours
ETA: 1/15/2023

**Comment by ExperimentsInHonesty** [comment]
@gstemmann Thank you for your work on this.  We are putting it in the icebox until we can move the wiki and then it will get added to the wiki in an appropriate section.  In the meantime its available here https://github.com/hackforla/website/wiki/Resources#web-developer
